src/dataflow/decode/mod.rs

```diff
@@ -116,24 +118,329 @@ where
+                                    key,
+                                    payload,
+                                    aux_num,
+                                    &mut session,
+                                    *cap.time(),
+                                );
+                            }
+                            Err(err) => {
+                                error!("{}", err);
+                            }
+                        }
+                    }
+                });
+                key_decoder_state.log_error_count();
+                value_decoder_state.log_error_count();
+            }
+        },
+    )
+}
+
+pub fn decode_kafka_upsert<G>(
+    stream: &Stream<G, ((Vec<u8>, Vec<u8>), Option<i64>)>,
+    value_encoding: DataEncoding,
+    key_encoding: DataEncoding,
+) -> Stream<G, ((Row, Option<Row>), Timestamp, Diff)>
+where
+    G: Scope<Timestamp = Timestamp>,
+{
+    let op_name = format!(
+        "{}-{}Decode",
+        key_encoding.op_name(),
+        value_encoding.op_name()
+    );
+    match (key_encoding, value_encoding) {
+        (DataEncoding::Text, DataEncoding::Avro(val_enc)) => decode_kafka_upsert_inner(
+            stream,
+            TextDecoderState,
+            avro::AvroDecoderState::new(
+                &val_enc.value_schema,
+                val_enc.schema_registry_url,
+                interchange::avro::EnvelopeType::Upsert,
+            ),
+            &op_name,
+        ),
+        (DataEncoding::Avro(key_enc), DataEncoding::Avro(val_enc)) => decode_kafka_upsert_inner(
+            stream,
+            avro::AvroDecoderState::new(
+                &key_enc.value_schema,
+                key_enc.schema_registry_url,
+                interchange::avro::EnvelopeType::None,
+            ),
+            avro::AvroDecoderState::new(
+                &val_enc.value_schema,
+                val_enc.schema_registry_url,
+                interchange::avro::EnvelopeType::Upsert,
+            ),
+            &op_name,
+        ),
+        _ => unreachable!(),
+    }
+}
+
+fn decode_kafka_inner<G, V: 'static>(
+    stream: &Stream<G, ((Vec<u8>, Vec<u8>), Option<i64>)>,
+    mut value_decoder_state: V,
+    op_name: &str,
+) -> Stream<G, (Row, Timestamp, Diff)>
+where
+    G: Scope<Timestamp = Timestamp>,
+    V: DecoderState,
+{
+    stream.unary(
+        Exchange::new(|x: &((_, Vec<u8>), _)| (x.0).1.hashed()),
+        &op_name,
+        move |_, _| {
+            move |input, output| {
+                value_decoder_state.reset_event_count();
+                input.for_each(|cap, data| {
+                    let mut session = output.session(&cap);
+                    for ((_, payload), aux_num) in data.iter() {
+                        value_decoder_state.give_value(payload, aux_num, &mut session, *cap.time());
+                    }
+                });
+                value_decoder_state.log_error_count();
+            }
+        },
+    )
+}
+
+pub fn decode_kafka<G>(
```

Comment:
An alternative to this method existing is to have in src/dataflow/source/kafka.rs have two methods to create the regular Row stream and the (Row, Option<Row>) stream, but that seemed more annoying, all the more given how we plan in the future to expose per-Row metadata anyway.